lib/dal/src/action/prototype.rs

```diff
@@ -362,10 +451,56 @@ impl ActionPrototype {
         Ok(action_prototype_ids)
     }
 
+    pub async fn parentage(
+        ctx: &DalContext,
+        id: ActionPrototypeId,
+    ) -> ActionPrototypeResult<ActionPrototypeParent> {
+        if let Some(schema_ids) = Self::schema_ids(ctx, id).await? {
+            Ok(ActionPrototypeParent::Schemas(schema_ids))
+        } else if let Some(schema_variant_id) = Self::schema_variant_id(ctx, id).await? {
+            Ok(ActionPrototypeParent::SchemaVariant(schema_variant_id))
+        } else {
+            Err(ActionPrototypeError::Orphaned(id))
+        }
+    }
+
+    /// If this Action Prototype is parented by one or more Schema(s), returns the Schema ids.
+    pub async fn schema_ids(
+        ctx: &DalContext,
+        id: ActionPrototypeId,
+    ) -> ActionPrototypeResult<Option<Vec<SchemaId>>> {
+        let workspace_snapshot = ctx.workspace_snapshot()?;
+        let mut result = vec![];
+
+        // Schemas have an outgoing edge to the action prototype, which have an incoming edge from the schema
+        let maybe_schemas = workspace_snapshot
+            .incoming_sources_for_edge_weight_kind(id, EdgeWeightKindDiscriminants::ActionPrototype)
+            .await?;
+
+        for maybe_schema_id in maybe_schemas {
+            if let NodeWeight::Content(content_weight) = ctx
+                .workspace_snapshot()?
+                .get_node_weight(maybe_schema_id)
+                .await?
+            {
+                if ContentAddressDiscriminants::Schema == content_weight.content_address().into() {
+                    result.push(maybe_schema_id.into());
+                }
+            }
+        }
+
+        if result.is_empty() {
+            Ok(None)
+        } else {
+            Ok(Some(result))
+        }
+    }
+
+    /// If this Action Prototype is parented by a SchemaVariant, returns the SchemaVariant Id.
```

Comment:
Thinking out loud here: might be less future confusion if schema_variant_id() and schema_ids didn't exist, and you had to call `parentage()` for everything. Calls to schema_variant_id() would look like

```rust
let ActionPrototypeParent::SchemaVariant(schema_variant_id) = ActionPrototype::parentage(ctx, ap_id) else { ... };
```

Not needed to change before checkin, what you have will work! But this seems like it'll be roughly the same amount of work for the caller (who has to check for `Some()` anyway, and exposes them to the real reason for the `Option` (the fact that prototypes can have other parents).